design/vela-core/definition-versioning.md

```diff
@@ -0,0 +1,296 @@
+- If both the annotations  `definitionrevision.oam.dev/name` and `definitionrevision.oam.dev/version` are present, there is no clear way to decide the Revision behavior.
+
+
+## Proposal 2
+
+This proposal targets the following major changes:
+
+- Add a new API version(`v2Beta1`) for ComponentDefinition CRD to support the new versioning model. Current Components should continue to work as is.
+- Use Semantic versioning to identify new versions of a Component.
+- Use the [Kubernetes API versioning model](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/) for managing the Component versions, which is to store an array of all version specifications. Each individual element in the array `spec.versions`  will also include a new field called `version` to indicate a specific Component version.
+- Kubevela Application users can refer to a particular version of a Component like `component-name@component-version`. The `version` must be a valid Semantic version without the prefix `v` for example, `1.1.0`, `1.2.0-rc.0`.
+
+#### ComponentDefinition CRD
+```
+    apiVersion: apiextensions.k8s.io/v1
+    kind: CustomResourceDefinition
+    metadata:
+      annotations:
+        cert-manager.io/inject-ca-from: vela-system/kubevela-vela-core-root-cert
+        controller-gen.kubebuilder.io/version: v0.11.4
+      name: componentdefinitions.core.oam.dev
+    spec:
+      conversion:
+        strategy: Webhook
+        webhook:
+          clientConfig:
+            service:
+              name: vela-core-webhook
+              namespace: vela-system
+              path: /convert
+              port: 443
+          conversionReviewVersions:
+          - v1beta1
+          - v2Beta1
+      group: core.oam.dev
+      names: ...
+      scope: Namespaced
+      versions:
+      - additionalPrinterColumns:
+        - jsonPath: .spec.workload.definition.kind ...
+        - jsonPath: .metadata.annotations.definition\.oam\.dev/ description ...
+        name: v1beta1
+        schema: ...
+        served: true
+        storage: true
+        subresources:
+          status: {}
+      - additionalPrinterColumns:
+        - jsonPath: .spec.workload.definition.kind ...
+        - jsonPath: .metadata.annotations.definition\.oam\.dev/description ...
+        name: v2Beta1
+        schema: ...
+        served: true
+        storage: false
+        subresources:
+          status: {}
+
+```
+
+#### ComponentDefinition v2Beta1 Object
+
+```
+    apiVersion: core.oam.dev/v2Beta1
+    kind: ComponentDefinition
+    metadata:
+      annotations: {}
+      name: my-component
+      namespace: vela-system
+    spec:
+      versions:
+      - version: "2"
+        schematic: ...
+        workload: ...
+      - version: "3.4"
+        schematic: ...
+        workload: ...
```

Comment:
Why combine versions in a single ComponentDefinition? Have you considered a more minimal CRD change to set a single `spec.version` property? Such a small change would perhaps justify using CRD version `v1beta2` instead of a major `v2beta1` version.

Reply:
Good point. Also having `versions` is more consistent with `CustomResourceDefinition` behavior, which also has `versions`.